source/opt/ir_builder.cpp

```diff
@@ -0,0 +1,117 @@
+//
+// THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+#include "ir_builder.h"
+
+#include <cassert>
+
+#include "reflect.h"
+
+namespace spvtools {
+namespace opt {
+namespace ir {
+
+void IrBuilder::AddInstruction(const spv_parsed_instruction_t* inst) {
+  const auto opcode = static_cast<SpvOp>(inst->opcode);
+  if (IsDebugLineInst(opcode)) {
+    dbg_line_info_.push_back(Instruction(*inst));
+    return;
+  }
+
+  Instruction spv_inst(*inst, std::move(dbg_line_info_));
+  dbg_line_info_.clear();
+  // Handle function and basic block boundaries first, then normal
+  // instructions.
+  if (opcode == SpvOpFunction) {
+    assert(function_ == nullptr);
+    assert(block_ == nullptr);
+    function_.reset(new Function(std::move(spv_inst)));
+  } else if (opcode == SpvOpFunctionEnd) {
+    assert(function_ != nullptr);
+    assert(block_ == nullptr);
+    module_->AddFunction(std::move(*function_.release()));
+    function_ = nullptr;
+  } else if (opcode == SpvOpLabel) {
+    assert(function_ != nullptr);
+    assert(block_ == nullptr);
+    block_.reset(new BasicBlock(std::move(spv_inst)));
+  } else if (IsTerminatorInst(opcode)) {
+    assert(function_ != nullptr);
+    assert(block_ != nullptr);
+    block_->AddInstruction(std::move(spv_inst));
+    function_->AddBasicBlock(std::move(*block_.release()));
+    block_ = nullptr;
+  } else {
+    if (function_ == nullptr) {  // Outside function definition
+      assert(block_ == nullptr);
+      if (opcode == SpvOpCapability) {
+        module_->AddCapability(std::move(spv_inst));
+      } else if (opcode == SpvOpExtension) {
+        module_->AddExtension(std::move(spv_inst));
+      } else if (opcode == SpvOpExtInstImport) {
+        module_->AddExtInstSet(std::move(spv_inst));
+      } else if (opcode == SpvOpMemoryModel) {
+        module_->SetMemoryModel(std::move(spv_inst));
+      } else if (opcode == SpvOpEntryPoint) {
+        module_->AddEntryPoint(std::move(spv_inst));
+      } else if (opcode == SpvOpExecutionMode) {
+        module_->AddExecutionMode(std::move(spv_inst));
+      } else if (IsDebugInst(opcode)) {
+        module_->AddDebugInst(std::move(spv_inst));
+      } else if (IsAnnotationInst(opcode)) {
+        module_->AddAnnotationInst(std::move(spv_inst));
+      } else if (IsTypeInst(opcode)) {
+        module_->AddType(std::move(spv_inst));
+      } else if (IsConstantInst(opcode)) {
+        module_->AddConstant(std::move(spv_inst));
+      } else if (opcode == SpvOpVariable) {
+        module_->AddVariable(std::move(spv_inst));
```

Comment:
Good point. Revised. OpVariables for local variables are put in the basic blocks.